File: packages/db/src/entities/Tag.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  Index,
  CreateDateColumn,
  UpdateDateColumn,
} from 'typeorm';

@Entity('tags')
@Index(['disciplineId', 'slug'], { unique: true })
export class Tag {
  @PrimaryGeneratedColumn('increment')
  id!: number;

  @Index()
  @Column({ type: 'bigint', nullable: false })
  disciplineId!: number;

  @Column({ type: 'varchar', length: 255, nullable: false })
  name!: string;

  @Column({ type: 'varchar', length: 255, nullable: false })
  slug!: string;

  @Column({ type: 'text', nullable: true })
  description!: string | null;

  @Column({ type: 'varchar', length: 50, nullable: true })
  color!: string | null; // hex color for UI display

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;
}

File: backend/internal/model/tag.go
package model

import "time"

type Tag struct {
	ID           string    `json:"id" firestore:"-"`
	DisciplineID string    `json:"disciplineId" firestore:"disciplineId"`
	Name         string    `json:"name" firestore:"name"`
	Slug         string    `json:"slug" firestore:"slug"`
	Description  string    `json:"description" firestore:"description"`
	Color        *string   `json:"color" firestore:"color,omitempty"`
	OwnerUID     string    `json:"ownerUid" firestore:"ownerUid"`
	CreatedAt    time.Time `json:"createdAt" firestore:"createdAt"`
	UpdatedAt    time.Time `json:"updatedAt" firestore:"updatedAt"`
}

type CreateTagRequest struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Color       *string `json:"color"`
}

type UpdateTagRequest struct {
	Name        *string `json:"name"`
	Description *string `json:"description"`
	Color       *string `json:"color"`
}
